// VaultSync service requests and responses

import "maidsafe/common/maidsafe.proto";

package maidsafe.protobuf;

message VaultAccountSet {
  message VaultAccount {
    required bytes  pmid = 1;
    required uint64 offered = 2;
    required uint64 vault_used = 3;
    required uint64 account_used = 4;
    repeated bytes alerts = 5;
  }
  repeated VaultAccount vault_account = 1;
}

message ChunkInfoMap {
  message VaultChunkInfo {
    message WaitingListEntry{
      required bytes pmid = 1;
      required uint32 creation_time = 2;
      required bool storing_done = 3;
      required bool payments_done = 4;
      required int32 requested_payments = 5;
    }
    message WatchListEntry{
      required bytes pmid = 1;
      required bool can_delete = 2;
    }
    message ReferenceListEntry{
      required bytes pmid = 1;
      required uint32 last_seen = 2;
    }
    required bytes chunk_name = 1;
    repeated WaitingListEntry waiting_list_entry = 2;
    repeated WatchListEntry watch_list_entry = 3;
    repeated ReferenceListEntry reference_list_entry = 4;
    required uint64 watcher_count = 5;
    required uint64 watcher_checksum = 6;
    required uint64 chunk_size = 7;
  }
  repeated VaultChunkInfo vault_chunk_info = 1;
}

message VaultBufferMap {
  message VaultBuffer {
    required bytes buffer_name = 1;
    repeated GenericPacket owner_info = 2;
  }
  repeated VaultBuffer vault_buffer = 1;
}


message GetSyncDataRequest {
  required bytes pmid = 1;
  required bytes public_key = 2;
  required bytes public_key_signature = 3;
  required bytes request_signature = 4;
}

message GetSyncDataResponse {
  required bool result = 1;
  optional VaultAccountSet vault_account_set = 2;
  optional ChunkInfoMap chunk_info_map = 3;
  optional VaultBufferMap vault_buffer_map = 4;
}

message GetAccountRequest {
  required bytes account_pmid = 1;  // PMID of account being queried
  required bytes pmid = 2;
  required bytes public_key = 3;
  required bytes public_key_signature = 4;
  required bytes request_signature = 5;
}

message GetAccountResponse {
  required bool result = 1;
  optional VaultAccountSet.VaultAccount vault_account = 2;
}

message GetChunkInfoRequest {
  required bytes chunkname = 1;
  required bytes pmid = 2;
  required bytes public_key = 3;
  required bytes public_key_signature = 4;
  required bytes request_signature = 5;
}

message GetChunkInfoResponse {
  required bool result = 1;
  optional ChunkInfoMap.VaultChunkInfo vault_chunk_info = 2;
}

message GetBufferRequest {
  required bytes buffer_name = 1;
  required bytes pmid = 2;
  required bytes public_key = 3;
  required bytes public_key_signature = 4;
  required bytes request_signature = 5;
}

message GetBufferResponse {
  required bool result = 1;
  optional VaultBufferMap.VaultBuffer vault_buffer = 2;
}